### ISSUE TITLE
Crash in `null` search in array fields in dsl queries

### ISSUE THREAD

**Issue by oruchreis** [Restream/reindexer#88]
Hi,
Reindexer crashes on this dsl query:
```json
{
  "namespace": "TestNs",
  "type": "select",
  "filters": [
    {
      "field": "ArrayField",
      "cond": "SET",
      "value": [1,2,null,3,4],
    }
  ],
}
```

### Actual:
As far as I can see from the log, in the dsl query, it is stated that the array field is `int` and `null` cannot be searched, but since this exception is not handled, the reindexer crashes. By the way `ArrayField` is marked as `sparse`.
```log
/src/cpp_src/core/keyvalue/variant.cc:67: failed assertion 'got.Is<T>()':
Expected value 'int', but got 'null'*** Current query dump ***
Query:    SELECT * FROM TestNs WHERE ArrayField IN (1,null,2,3,4)
NS state: Optimization completed
NS.locker state:  regular
NS.strHolder state: [
memstat = 0
holds indexes = false
]
Explain:  {"total_us":0,"preselect_us":0,"prepare_us":15,"indexes_us":0,"postprocess_us":0,"loop_us":0,"general_sort_us":0,"sort_index":"","sort_by_uncommitted_index":false,"selectors":[]}
[C11/12 07:04:34.637 2736] *** Backtrace on signal: 6 ***
 
[C11/12 07:04:34.637 2736] *** No additional info from crash query tracker ***
 
[C11/12 07:04:34.637 2736] Assertion message: <empty>
 
[C11/12 07:04:35.618 2736] Signal 6 backtrace (libunwind):
#1 0x007fb885d26ebc <unresolved_function> + 140430495870652
#2 0x007fb885cd7fb2 gsignal + 18
#3 0x007fb885cc2472 abort + 211
#4 0x005646f0cbd0ad <unresolved_function> (cpp_src/core/keyvalue/variant.cc:67)
#5 0x005646f1262d23 std::_Function_handler<bool (reindexer::SelectKeyResult&, unsigned long&), reindexer::IndexUnordered<reindexer::unordered_number_map<int, reindexer::KeyEntry<reindexer::IdSetPlain> > >::SelectKey(reindexer::VariantArray const&, CondType, unsigned int, reindexer::Index::SelectOpts, reindexer::intrusive_ptr<reindexer::BaseFunctionCtx> const&, reindexer::RdxContext const&)::{lambda(reindexer::SelectKeyResult&, unsigned long&)#1}>::_M_invoke(std::_Any_data const&, reindexer::SelectKeyResult&, unsigned long&) (/usr/include/c++/12/bits/std_function.h:290)
#6 0x005646f12c244d reindexer::IndexUnordered<reindexer::unordered_number_map<int, reindexer::KeyEntry<reindexer::IdSetPlain> > >::tryIdsetCache(reindexer::VariantArray const&, CondType, unsigned int, std::function<bool (reindexer::SelectKeyResult&, unsigned long&)> const&, reindexer::SelectKeyResult&) (cpp_src/core/index/indexunordered.cc:282)
#7 0x005646f12c2b6c reindexer::IndexUnordered<reindexer::unordered_number_map<int, reindexer::KeyEntry<reindexer::IdSetPlain> > >::SelectKey(reindexer::VariantArray const&, CondType, unsigned int, reindexer::Index::SelectOpts, reindexer::intrusive_ptr<reindexer::BaseFunctionCtx> const&, reindexer::RdxContext const&) (cpp_src/core/index/indexunordered.cc:346)
#8 0x005646f0e46f14 reindexer::SelectIteratorContainer::processQueryEntry(reindexer::QueryEntry const&, bool, reindexer::NamespaceImpl const&, unsigned int, bool, reindexer::FtSortType, reindexer::intrusive_ptr<reindexer::SelectFunction>&, bool&, bool&, reindexer::intrusive_ptr<reindexer::FtCtx>&, reindexer::QueryPreprocessor&, reindexer::RdxContext const&) (cpp_src/core/nsselecter/selectiteratorcontainer.cc:323)
#9 0x005646f0e4f626 reindexer::SelectIteratorContainer::prepareIteratorsForSelectLoop(reindexer::QueryPreprocessor&, unsigned long, unsigned long, unsigned int, bool, reindexer::FtSortType, reindexer::NamespaceImpl const&, reindexer::intrusive_ptr<reindexer::SelectFunction>&, reindexer::intrusive_ptr<reindexer::FtCtx>&, reindexer::RdxContext const&) (cpp_src/core/nsselecter/selectiteratorcontainer.cc:532)
#10 0x005646f0e4fd21 reindexer::SelectIteratorContainer::PrepareIteratorsForSelectLoop(reindexer::QueryPreprocessor&, unsigned int, bool, reindexer::FtSortType, reindexer::NamespaceImpl const&, reindexer::intrusive_ptr<reindexer::SelectFunction>&, reindexer::intrusive_ptr<reindexer::FtCtx>&, reindexer::RdxContext const&) (cpp_src/core/nsselecter/selectiteratorcontainer.cc:519)
#11 0x005646f0e0828f void reindexer::NsSelecter::operator()<void>(reindexer::QueryResults&, reindexer::SelectCtxWithJoinPreSelect<void>&, reindexer::RdxContext const&) (cpp_src/core/nsselecter/nsselecter.cc:194)
#12 0x005646f0cff74c void reindexer::NamespaceImpl::Select<void>(reindexer::QueryResults&, reindexer::SelectCtxWithJoinPreSelect<void>&, reindexer::RdxContext const&) (cpp_src/core/namespace/namespaceimpl.cc:2423)
#13 0x005646f0f9ef68 void reindexer::RxSelector::DoSelect<reindexer::RdxContext const, reindexer::long_actions::QueryEnum2Type<(QueryType)0> >(reindexer::Query const&, reindexer::QueryResults&, reindexer::RxSelector::NsLocker<reindexer::RdxContext const>&, reindexer::SelectFunctionsHolder&, reindexer::RdxContext const&, reindexer::QueryStatCalculator<reindexer::long_actions::QueryEnum2Type<(QueryType)0>, reindexer::long_actions::Logger>&) (cpp_src/core/reindexer_impl/rx_selector.cc:107)
#14 0x005646f0f77cd2 reindexer::ReindexerImpl::Select(reindexer::Query const&, reindexer::QueryResults&, reindexer::InternalRdxContext const&) (cpp_src/core/reindexer_impl/reindexerimpl.cc:955)
#15 0x005646f0f5c343 reindexer::Reindexer::Select(reindexer::Query const&, reindexer::QueryResults&) (cpp_src/core/reindexer.cc:58)
#16 0x007fb8853aa640 reindexer::grpc::ReindexerService::executeQuery(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, reindexer::grpc::Query const&, QueryType, reindexer::QueryResults&) (cpp_src/server/grpc/reindexerservice.cc:653)
#17 0x007fb8853af455 reindexer::grpc::ReindexerService::Select(grpc::ServerContext*, reindexer::grpc::SelectRequest const*, grpc::ServerWriter<reindexer::grpc::QueryResultsResponse>*) (cpp_src/server/grpc/reindexerservice.cc:690)
#18 0x007fb88534b7fc <unresolved_function> (/usr/include/c++/12/bits/std_function.h:292)
#19 0x007fb8853757fc grpc::Status grpc::internal::CatchingFunctionHandler<grpc::internal::ServerStreamingHandler<reindexer::grpc::Reindexer::Service, reindexer::grpc::SelectRequest, reindexer::grpc::QueryResultsResponse>::RunHandler(grpc::internal::MethodHandler::HandlerParameter const&)::{lambda()#1}>(grpc::internal::ServerStreamingHandler<reindexer::grpc::Reindexer::Service, reindexer::grpc::SelectRequest, reindexer::grpc::QueryResultsResponse>::RunHandler(grpc::internal::MethodHandler::HandlerParameter const&)::{lambda()#1}&&) (/usr/include/grpcpp/support/method_handler.h:50)
#20 0x007fb88537ec23 grpc::internal::ServerStreamingHandler<reindexer::grpc::Reindexer::Service, reindexer::grpc::SelectRequest, reindexer::grpc::QueryResultsResponse>::RunHandler(grpc::internal::MethodHandler::HandlerParameter const&) (/usr/include/grpcpp/support/method_handler.h:202)
#21 0x007fb884f62a82 grpc::Server::SyncRequest::ContinueRunAfterInterception() + 386
#22 0x007fb884f6955b grpc::ThreadManager::MainWorkLoop() + 235
#23 0x007fb884f6967c grpc::ThreadManager::WorkerThread::Run() + 12
#24 0x007fb8847e3ef3 <unresolved_function> + 140430473576179
#25 0x007fb885d251c4 <unresolved_function> + 140430495863236
#26 0x007fb885da585c <unresolved_function> + 140430496389212
```

### Expected:
The behaviour we expect is that Reindexer returns the error `Expected value 'int', but got 'null'`, but does not crash. 

### Expected 2:
Another expected behaviour is that since `ArrayField` is actually sparse, we expect the array elements to be nullable. In the actual behavior, we have to write weird queries like this to search `null` in the array:
```json
{
  "filters": [
    {
      "filters": [
        {
          "field": "ArrayField",
          "cond": "EMPTY"  /*the array contains null or array itself is null*/
        },
        {
          "field": "ArrayField",
          "cond": "SET",
          "value": [1,2,3,4]
        }
      ],
      "op": "AND"
    },
    {
      "field": "ArrayField",
      "cond": "ANY"        /*the array must not be null*/
    }
  ]
}
```
However, since the `ArrayField` is actually `sparse`, we should be able to search for `null` values in `SET` and `ALLSET` like dsl query at top:
```json
 "filters": [
    {
      "field": "ArrayField",
      "cond": "SET",
      "value": [1,2,null,3,4],
      "op": "OR"
    },
    {
      "field": "ArrayField",
      "cond": "ALLSET",
      "value": [null,3,4],
    },
  ],
```

**Comment by MadSchemas** [comment]
Hi,
Thank you for the bugreport. The second suggested behavior seems more correct to me. We will check if it does not conflicts with some other logic and will try to implement it in one of the next releases (in January, probably)

**Comment by MadSchemas** [comment]
The crash was fixed in [v4.19.0](https://github.com/Restream/reindexer/releases/tag/v4.19.0). Now reindexer will return error in this scenario.

The second approach with explicit 'null' in values set will be implemented in one of the next releases with some more fixes, which are required for consistent behavior of null-values handling in indexed/non-indexed comparators and indexing structures

**Comment by oruchreis** [comment]
Thanks for 3.31, it seems it has been fixed now.
